v2/sender.go

```diff
@@ -154,7 +164,10 @@ func (d *Sender) SendMessageBatch(ctx context.Context, messages []*azservicebus.
 	errChan := make(chan error)
 
 	go func() {
-		if err := d.sbSender.SendMessageBatch(ctx, batch, nil); err != nil {
+		d.mu.RLock()
+		err := d.sbSender.SendMessageBatch(ctx, batch, nil)
+		d.mu.RUnlock()
```

Comment:
always use defer
if SendMessageBatch panics, you won't be unlocking the mutex, and we're in trouble.
Moving the `lock/defer unlock` inside the `go func()` already scopes it better than before.
You could wrap even further in a func to avoid including the error handling in the critical section.